File: Problem1_Hashmap.cpp
// Time Complexity : O(number of keys/number of buckets)
// Space Complexity : O(number of keys + number of buckets)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : I struggled with initializations and understanding how to work with pairs and use of ref address. Can you please tell me how to become stronger in these concepts and also how to improve my speed?

class MyHashMap {
    vector<list<pair<int,int>>>tables;
    int table_size = 10000;
    
public:
    /** Initialize your data structure here. */
    MyHashMap() {
        tables.resize(table_size);
    }
    
    int hash_function(int key){
        return key%table_size;
    }
    
    /** value will always be non-negative. */
    void put(int key, int value) {
        int i = hash_function(key);
        auto &list = tables[i];
        for(auto &val : list){
            if(val.first == key){
                cout<<"Replacing the value as key exists\n";
                val.second = value;
                return;
            }
        }
        list.emplace_back(key, value);
    }
    
    /** Returns the value to which the specified key is mapped, or -1 if this map contains no mapping for the key */
    int get(int key) {
        int i = hash_function(key);
        auto &list = tables[i];
        if(list.empty()){
            return -1;
        }
        for(auto &val:list){
            if(val.first ==key){
                return val.second;
            }
        }
        return -1;
    }
    
    /** Removes the mapping of the specified value key if this map contains a mapping for the key */
    void remove(int key) {
        int i = hash_function(key);
        auto &list = tables[i];
        auto n = begin(list);
        list.remove_if([key](auto n){return n.first==key;});
    }
};

/**
 * Your MyHashMap object will be instantiated and called as such:
 * MyHashMap* obj = new MyHashMap();
 * obj->put(key,value);
 * int param_2 = obj->get(key);
 * obj->remove(key);
 */

File: Problem_1_Hashmap.py


class ListNode(object):

    def __init__(self,key,val):
        """
        Initialize your data structure here.
        """
        #Using the linked list we look at the most recent added node and move to the next until find the correct node
        self.pair = (key,val)
        self.next = None


class MyHashMap:

    def __init__(self):
        self.m = 1000;
        self.h = [None]*self.m


    def put(self, key, value):
        """
        value will always be non-negative.
        :type key: int
        :type value: int
        :rtype: None
        """
        index = key % self.m
        if self.h[index] == None:
            self.h[index] = ListNode(key,value)
        else:
            cur = self.h[index]
            while True:
                #if current key is the index then update the value in the key at that index
                if cur.pair[0] == key:
                    cur.pair = (key,value)
                    return
                if cur.next == None:
                    break
                cur = cur.next
            cur.next = ListNode(key,value)



    def get(self, key):
        """
        Returns the value to which the specified key is mapped, or -1 if this map contains no mapping for the key
        :type key: int
        :rtype: int
        """
        index = key % self.m
        cur = self.h[index]
        while cur:
            if cur.pair[0] == key:
                return cur.pair[1]
            else:
                cur = cur.next
        return -1
        

    def remove(self, key):
        """
        Removes the mapping of the specified value key if this map contains a mapping for the key
        :type key: int
        :rtype: None
        """
        index = key % self.m
        cur = prev = self.h[index]
        if not cur:
            return
        if cur.pair[0] == key:
            self.h[index] = cur.next
        else:
            cur = cur.next
            while cur:
                if cur.pair[0] == key:
                    prev.next = cur.next
                    break
                else:
                    cur,prev = cur.next,prev.next

#Used array as direct table and linked list for key,value as pair. The time complexity for this hashmap for each function turns to be O(1).

# Your MyHashMap object will be instantiated and called as such:
#obj = MyHashMap()
#obj.put(key,value)
#param_2 = obj.get(key)
#obj.remove(key)